src/BCReader.jl

```diff
@@ -0,0 +1,309 @@
+#=
+    BCReader
+
+This module coordinates reading of boundary conditions from NetCDF files,
+as well as regridding calls and temporal interpolations from
+monthly to daily intervals.
+=#
+module BCReader
+
+using ..Utilities, ..Regridder, ..CallbackManager
+using ClimaCore: Fields
+using ClimaComms
+using Dates
+using JLD2
+
+export BCFileInfo, bcfile_info_init, update_midmonth_data!, next_date_in_file, interpolate_midmonth_to_daily
+
+
+"""
+    BCFileInfo
+
+Stores information specific to each boundary condition from a file and each variable.
+
+# Inputs
+    FT::F                           # float type
+    bcfile_dir::b                   # directory of the BC file
+    comms_ctx::X                    # communication context used for MPI
+    hd_outfile_root::S              # filename root for regridded data 
+    varname::V                      # name of the variable
+    all_dates::D                    # vector of all dates contained in the original data file
+    segment_idx::Vector{Int}        # index of the monthly data in the file
+    segment_idx0::Vector{Int}       # `segment_idx` of the file data that is closest to date0
+    monthly_fields::C               # tuple of the two monthly fields, that will be used for the daily interpolation
+    segment_length::Vector{Int}     # length of each month segment (used in the daily interpolation)
+    scaling_function::O             # function that scales, offsets or transforms the raw variable
+    interpolate_daily::Bool         # switch to trigger daily interpolation
+    land_mask::M                    # mask with 1 = land, 0 = ocean / sea-ice
+"""
+struct BCFileInfo{F, B, X, S, V, D, C, O, M}
+    FT::F
+    bcfile_dir::B
+    comms_ctx::X
+    hd_outfile_root::S
+    varname::V
+    all_dates::D
+    segment_idx::Vector{Int}
+    segment_idx0::Vector{Int}
+    monthly_fields::C
+    segment_length::Vector{Int}
+    scaling_function::O
+    interpolate_daily::Bool
+    land_mask::M
+end
```

Comment:
```suggestion
struct BCFileInfo{FT, B, X, S, V, D, C, O, M}
    bcfile_dir::B
    comms_ctx::X
    hd_outfile_root::S
    varname::V
    all_dates::D
    segment_idx::Vector{Int}
    segment_idx0::Vector{Int}
    monthly_fields::C
    segment_length::Vector{Int}
    scaling_function::O
    interpolate_daily::Bool
    land_mask::M
end
float_type(::BCFileInfo{FT}) where {FT} = FT
```
Storing types in struct fields is actually a performance issue-- the `typeof(Float64)` is a `DataType`, and the compiler cannot optimize fields of this type.

By chance, was this pattern adopted from somewhere else?

Reply:
It was - I'll open a little PR to address all the places in the code this style is used

Reply:
It seems like with the new style, no default constructors are created for the struct. I created a MWE 
`struct Foo{FT, S}
       val::S
       end`, which then gives `methods(Foo) 0 methods for type constructor`.
Is there an easy way around this? I could write explicit constructors for all the structs, but they're pretty large.

Reply:
There are a few options, one is to make a generic-like outer constructor:
```
julia> struct Foo{FT, S} val::S end

julia> Foo{FT}(args...) where {FT} = Foo{FT, typeof.(args)...}(args...)

julia> Foo{Float64}(1)
Foo{Float64, Int64}(1)
```
Another option is basically that with the use of the `Base.@kwdef` for adding a keyword constructor